pointercrate-user/src/auth/get.rs

```diff
@@ -46,9 +60,117 @@ impl AuthenticatedUser {
         Ok(user)
     }
 
+    pub async fn oauth2_callback(code: &str, existing_id: Option<i32>, connection: &mut PgConnection) -> Result<AuthenticatedUser> {
+        info!("We are expected to perform Google OAuth2 authentication");
+
+        let client = reqwest::Client::new();
+
+        let response = client
+            .post("https://oauth2.googleapis.com/token")
+            .form(&[
+                ("code", code),
+                ("client_id", &config::google_client_id()),
+                ("client_secret", &config::google_client_secret()),
+                ("redirect_uri", &config::google_redirect_uri()),
+                ("grant_type", "authorization_code"),
+            ])
+            .send()
+            .await
+            .map_err(|_| CoreError::Unauthorized)?;
+
+        let response: GoogleTokenResponse = response.json().await.map_err(|_| CoreError::Unauthorized)?;
+
+        // We can safely disable all validation here, as Google recommends to not
+        // validate a fresh token, as it is guaranteed to be valid.
+        //
+        // https://developers.google.com/identity/openid-connect/openid-connect#obtainuserinfo
+        let key = DecodingKey::from_secret(&[]);
+        let mut validation = Validation::new(Algorithm::HS256);
+        validation.insecure_disable_signature_validation();
+        validation.validate_exp = false;
+        validation.validate_nbf = false;
+        validation.validate_aud = false;
+
+        let user_info = jsonwebtoken::decode::<GoogleUserInfo>(&response.id_token, &key, &validation)
+            .map_err(|_| CoreError::Unauthorized)?
+            .claims;
+
+        match User::by_google_account(&user_info.id, connection).await {
+            Ok(user) => Ok(AuthenticatedUser {
+                user,
+                password_hash: None,
+                google_account_id: Some(user_info.id),
+                email_address: Some(user_info.email),
+            }),
+            Err(UserError::UserNotFoundGoogleAccount { .. }) => {
```

Comment:
Let's factor this entire match arm into some more functions:
- one for updating the google account id of an existing user (can probably go into auth/patch.rs)
- one for creating a new user account from a google account (maybe into auth/post.rs?)